backend/src/schema/resolvers/posts.spec.js

```diff
@@ -1,458 +1,524 @@
-      await expect(client.request(updatePostMutation, updatePostVariables)).resolves.toEqual(
+      const expected = { data: { UpdatePost: { id: 'p9876', content: 'New content' } } }
+      await expect(mutate({ mutation: updatePostMutation, variables })).resolves.toMatchObject(
         expected,
       )
     })
 
-    describe('categories', () => {
-      it('allows a user to update other attributes without passing in categoryIds explicitly', async () => {
-        const expected = { UpdatePost: { id: postId, content: newContent } }
-        await expect(client.request(updatePostMutation, updatePostVariables)).resolves.toEqual(
+    describe('no new category ids provided for update', () => {
+      it('resolves and keeps current categories', async () => {
+        const expected = {
+          data: {
+            UpdatePost: {
+              id: 'p9876',
+              categories: expect.arrayContaining([{ id: 'cat9' }, { id: 'cat4' }, { id: 'cat15' }]),
+            },
+          },
+        }
+        await expect(mutate({ mutation: updatePostMutation, variables })).resolves.toMatchObject(
           expected,
         )
       })
+    })
 
-      it('allows a user to update the categories of a post', async () => {
-        updatePostVariables.categoryIds = ['cat27']
-        await client.request(updatePostMutation, updatePostVariables)
-        const expected = [{ id: 'cat27' }]
-        const postQueryWithCategoriesVariables = {
-          id: postId,
-        }
-        await expect(
-          client.request(postQueryWithCategories, postQueryWithCategoriesVariables),
-        ).resolves.toEqual({ Post: [{ categories: expect.arrayContaining(expected) }] })
+    describe('given category ids', () => {
+      beforeEach(() => {
+        variables = { ...variables, categoryIds: ['cat27'] }
       })
 
-      it('allows a maximum of three category for a successful update', async () => {
-        updatePostVariables.categoryIds = ['cat9', 'cat27', 'cat15', 'cat4']
-        await expect(client.request(updatePostMutation, updatePostVariables)).rejects.toThrow(
-          postSaveError,
+      it('updates categories of a post', async () => {
+        const expected = {
+          data: {
+            UpdatePost: {
+              id: 'p9876',
+              categories: expect.arrayContaining([{ id: 'cat27' }]),
+            },
+          },
+        }
+        await expect(mutate({ mutation: updatePostMutation, variables })).resolves.toMatchObject(
+          expected,
         )
       })
 
+      describe('more than 3 categories', () => {
+        beforeEach(() => {
+          variables = { ...variables, categoryIds: ['cat9', 'cat27', 'cat15', 'cat4'] }
+        })
+
+        it('allows a maximum of three category for a successful update', async () => {
+          const {
+            errors: [error],
+          } = await mutate({ mutation: updatePostMutation, variables })
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-09-03T10:46:28Z" title="Tuesday, September 3rd 2019, 12:46:28 pm +02:00">Sep 3, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -1,458 +1,524 @@
-import { GraphQLClient } from 'graphql-request'
 import { createTestClient } from 'apollo-server-testing'
 import Factory from '../../seed/factories'
-import { host, login, gql } from '../../jest/helpers'
-import { neode, getDriver } from '../../bootstrap/neo4j'
+import { gql } from '../../jest/helpers'
+import { neode as getNeode, getDriver } from '../../bootstrap/neo4j'
 import createServer from '../../server'
 
 const driver = getDriver()
 const factory = Factory()
-const instance = neode()
-
-let client
-let userParams
-let authorParams
-
-const postId = 'p3589'
-const postTitle = 'I am a title'
-const postContent = 'Some content'
-const oldTitle = 'Old title'
-const oldContent = 'Old content'
-const newTitle = 'New title'
-const newContent = 'New content'
-const postSaveError = 'You cannot save a post without at least one category or more than three'
-const categoryIds = ['cat9', 'cat4', 'cat15']
-let createPostVariables
+const neode = getNeode()
 
-const postQueryWithCategories = gql`
-  query($id: ID) {
-    Post(id: $id) {
-      categories {
-        id
-      }
-    }
-  }
-`
-const postQueryFilteredByCategory = gql`
-  query Post($filter: _PostFilter) {
-    Post(filter: $filter) {
-      title
-      id
-      categories {
-        id
-      }
-    }
-  }
-`
-const postCategoriesFilterParam = { categories_some: { id_in: categoryIds } }
-const postQueryFilteredByCategoryVariables = {
-  filter: postCategoriesFilterParam,
-}
+let query
+let mutate
+let authenticatedUser
+let user
+
+const categoryIds = ['cat9', 'cat4', 'cat15']
+let variables
 
 const createPostMutation = gql`
-  mutation($id: ID, $title: String!, $content: String!, $categoryIds: [ID]) {
-    CreatePost(id: $id, title: $title, content: $content, categoryIds: $categoryIds) {
+  mutation($id: ID, $title: String!, $content: String!, $language: String, $categoryIds: [ID]) {
+    CreatePost(
+      id: $id
+      title: $title
+      content: $content
+      language: $language
+      categoryIds: $categoryIds
+    ) {
       id
       title
       content
       slug
       disabled
       deleted
+      language
+      author {
+        name
+      }
     }
   }
 `
+
+beforeAll(() => {
+  const { server } = createServer({
+    context: () => {
+      return {
+        driver,
+        neode,
+        user: authenticatedUser,
+      }
+    },
+  })
+  query = createTestClient(server).query
+  mutate = createTestClient(server).mutate
+})
+
 beforeEach(async () => {
-  userParams = {
+  variables = {}
+  user = await factory.create('User', {
     id: 'u198',
     name: 'TestUser',
     email: 'test@example.org',
     password: '1234',
-  }
-  authorParams = {
-    id: 'u25',
-    email: 'author@example.org',
-    password: '1234',
-  }
-  await factory.create('User', userParams)
+  })
   await Promise.all([
-    instance.create('Category', {
+    neode.create('Category', {
       id: 'cat9',
       name: 'Democracy & Politics',
       icon: 'university',
     }),
-    instance.create('Category', {
+    neode.create('Category', {
       id: 'cat4',
       name: 'Environment & Nature',
       icon: 'tree',
     }),
-    instance.create('Category', {
+    neode.create('Category', {
       id: 'cat15',
       name: 'Consumption & Sustainability',
       icon: 'shopping-cart',
     }),
-    instance.create('Category', {
+    neode.create('Category', {
       id: 'cat27',
       name: 'Animal Protection',
       icon: 'paw',
     }),
   ])
-  createPostVariables = {
-    id: postId,
-    title: postTitle,
-    content: postContent,
-    categoryIds,
-  }
+  authenticatedUser = null
 })
 
 afterEach(async () => {
   await factory.cleanDatabase()
 })
 
+describe('Post', () => {
+  const postQuery = gql`
+    query Post($filter: _PostFilter) {
+      Post(filter: $filter) {
+        id
+        categories {
+          id
+        }
+      }
+    }
+  `
+
+  describe('can be filtered', () => {
+    it('by categories', async () => {
+      await Promise.all([
+        factory.create('Post', { id: 'p31', categoryIds: ['cat4'] }),
+        factory.create('Post', { id: 'p32', categoryIds: ['cat15'] }),
+        factory.create('Post', { id: 'p33', categoryIds: ['cat9'] }),
+      ])
+      const expected = {
+        data: {
+          Post: [
+            {
+              id: 'p33',
+              categories: [{ id: 'cat9' }],
+            },
+          ],
+        },
+      }
+      variables = { ...variables, filter: { categories_some: { id_in: ['cat9'] } } }
+      await expect(query({ query: postQuery, variables })).resolves.toMatchObject(expected)
+    })
+  })
+})
+
 describe('CreatePost', () => {
+  beforeEach(() => {
+    variables = {
+      ...variables,
+      id: 'p3589',
+      title: 'I am a title',
+      content: 'Some content',
+      categoryIds,
+    }
+  })
+
   describe('unauthenticated', () => {
     it('throws authorization error', async () => {
-      client = new GraphQLClient(host)
-      await expect(client.request(createPostMutation, createPostVariables)).rejects.toThrow(
-        'Not Authorised',
-      )
+      const { errors } = await mutate({ mutation: createPostMutation, variables })
+      expect(errors[0]).toHaveProperty('message', 'Not Authorised!')
     })
   })
 
   describe('authenticated', () => {
-    let headers
     beforeEach(async () => {
-      headers = await login(userParams)
-      client = new GraphQLClient(host, { headers })
+      authenticatedUser = await user.toJson()
     })
 
     it('creates a post', async () => {
+      const expected = { data: { CreatePost: { title: 'I am a title', content: 'Some content' } } }
+      await expect(mutate({ mutation: createPostMutation, variables })).resolves.toMatchObject(
+        expected,
+      )
+    })
+
+    it('assigns the authenticated user as author', async () => {
       const expected = {
-        CreatePost: {
-          title: postTitle,
-          content: postContent,
+        data: {
+          CreatePost: {
+            title: 'I am a title',
+            author: {
+              name: 'TestUser',
+            },
+          },
         },
       }
-      await expect(client.request(createPostMutation, createPostVariables)).resolves.toMatchObject(
+      await expect(mutate({ mutation: createPostMutation, variables })).resolves.toMatchObject(
         expected,
       )
     })
 
-    it('assigns the authenticated user as author', async () => {
-      await client.request(createPostMutation, createPostVariables)
-      const { User } = await client.request(
-        gql`
-          {
-            User(name: "TestUser") {
-              contributions {
-                title
-              }
-            }
-          }
-        `,
-        { headers },
+    it('`disabled` and `deleted` default to `false`', async () => {
+      const expected = { data: { CreatePost: { disabled: false, deleted: false } } }
+      await expect(mutate({ mutation: createPostMutation, variables })).resolves.toMatchObject(
+        expected,
       )
-      expect(User).toEqual([{ contributions: [{ title: postTitle }] }])
     })
 
-    describe('disabled and deleted', () => {
-      it('initially false', async () => {
-        const expected = { CreatePost: { disabled: false, deleted: false } }
-        await expect(
-          client.request(createPostMutation, createPostVariables),
-        ).resolves.toMatchObject(expected)
+    describe('language', () => {
+      beforeEach(() => {
+        variables = { ...variables, language: 'es' }
       })
-    })
 
-    describe('language', () => {
       it('allows a user to set the language of the post', async () => {
-        const createPostWithLanguageMutation = gql`
-          mutation($title: String!, $content: String!, $language: String, $categoryIds: [ID]) {
-            CreatePost(
-              title: $title
-              content: $content
-              language: $language
-              categoryIds: $categoryIds
-            ) {
-              language
-            }
-          }
-        `
-        const createPostWithLanguageVariables = {
-          title: postTitle,
-          content: postContent,
-          language: 'en',
-          categoryIds,
-        }
-        const expected = { CreatePost: { language: 'en' } }
-        await expect(
-          client.request(createPostWithLanguageMutation, createPostWithLanguageVariables),
-        ).resolves.toEqual(expect.objectContaining(expected))
+        const expected = { data: { CreatePost: { language: 'es' } } }
+        await expect(mutate({ mutation: createPostMutation, variables })).resolves.toMatchObject(
+          expected,
+        )
       })
     })
 
     describe('categories', () => {
-      it('throws an error if categoryIds is not an array', async () => {
-        createPostVariables.categoryIds = null
-        await expect(client.request(createPostMutation, createPostVariables)).rejects.toThrow(
-          postSaveError,
-        )
-      })
-
-      it('requires at least one category for successful creation', async () => {
-        createPostVariables.categoryIds = []
-        await expect(client.request(createPostMutation, createPostVariables)).rejects.toThrow(
-          postSaveError,
-        )
+      describe('null', () => {
+        beforeEach(() => {
+          variables = { ...variables, categoryIds: null }
+        })
+        it('throws UserInputError', async () => {
+          const {
+            errors: [error],
+          } = await mutate({ mutation: createPostMutation, variables })
+          expect(error).toHaveProperty(
+            'message',
+            'You cannot save a post without at least one category or more than three',
+          )
+        })
       })
 
-      it('allows a maximum of three category for successful update', async () => {
-        createPostVariables.categoryIds = ['cat9', 'cat27', 'cat15', 'cat4']
-        await expect(client.request(createPostMutation, createPostVariables)).rejects.toThrow(
-          postSaveError,
-        )
+      describe('empty', () => {
+        beforeEach(() => {
+          variables = { ...variables, categoryIds: [] }
+        })
+        it('throws UserInputError', async () => {
+          const {
+            errors: [error],
+          } = await mutate({ mutation: createPostMutation, variables })
+          expect(error).toHaveProperty(
+            'message',
+            'You cannot save a post without at least one category or more than three',
+          )
+        })
       })
 
-      it('allows a user to filter for posts by category', async () => {
-        await client.request(createPostMutation, createPostVariables)
-        const categoryIdsArray = [{ id: 'cat4' }, { id: 'cat15' }, { id: 'cat9' }]
-        const expected = {
-          Post: [
-            {
-              title: postTitle,
-              id: postId,
-              categories: expect.arrayContaining(categoryIdsArray),
-            },
-          ],
-        }
-        await expect(
-          client.request(postQueryFilteredByCategory, postQueryFilteredByCategoryVariables),
-        ).resolves.toEqual(expected)
+      describe('more than 3 items', () => {
+        beforeEach(() => {
+          variables = { ...variables, categoryIds: ['cat9', 'cat27', 'cat15', 'cat4'] }
+        })
+        it('throws UserInputError', async () => {
+          const {
+            errors: [error],
+          } = await mutate({ mutation: createPostMutation, variables })
+          expect(error).toHaveProperty(
+            'message',
+            'You cannot save a post without at least one category or more than three',
+          )
+        })
       })
     })
   })
 })
 
 describe('UpdatePost', () => {
-  let updatePostVariables
+  let author
   const updatePostMutation = gql`
     mutation($id: ID!, $title: String!, $content: String!, $categoryIds: [ID]) {
       UpdatePost(id: $id, title: $title, content: $content, categoryIds: $categoryIds) {
         id
         content
+        categories {
+          id
+        }
       }
     }
   `
   beforeEach(async () => {
-    const asAuthor = Factory()
-    await asAuthor.create('User', authorParams)
-    await asAuthor.authenticateAs(authorParams)
-    await asAuthor.create('Post', {
-      id: postId,
-      title: oldTitle,
-      content: oldContent,
+    author = await factory.create('User', { slug: 'the-author' })
+    await factory.create('Post', {
+      author,
+      id: 'p9876',
+      title: 'Old title',
+      content: 'Old content',
       categoryIds,
     })
 
-    updatePostVariables = {
-      id: postId,
-      title: newTitle,
-      content: newContent,
+    variables = {
+      ...variables,
+      id: 'p9876',
+      title: 'New title',
+      content: 'New content',
     }
   })
 
   describe('unauthenticated', () => {
     it('throws authorization error', async () => {
-      client = new GraphQLClient(host)
-      await expect(client.request(updatePostMutation, updatePostVariables)).rejects.toThrow(
-        'Not Authorised',
-      )
+      const { errors } = await mutate({ mutation: updatePostMutation, variables })
+      expect(errors[0]).toHaveProperty('message', 'Not Authorised!')
     })
   })
 
   describe('authenticated but not the author', () => {
-    let headers
     beforeEach(async () => {
-      headers = await login(userParams)
-      client = new GraphQLClient(host, { headers })
+      authenticatedUser = await user.toJson()
     })
 
     it('throws authorization error', async () => {
-      await expect(client.request(updatePostMutation, updatePostVariables)).rejects.toThrow(
-        'Not Authorised',
-      )
+      const { errors } = await mutate({ mutation: updatePostMutation, variables })
+      expect(errors[0]).toHaveProperty('message', 'Not Authorised!')
     })
   })
 
   describe('authenticated as author', () => {
-    let headers
     beforeEach(async () => {
-      headers = await login(authorParams)
-      client = new GraphQLClient(host, { headers })
+      authenticatedUser = await author.toJson()
     })
 
     it('updates a post', async () => {
-      updatePostVariables.categoryIds = ['cat27']
-      const expected = { UpdatePost: { id: postId, content: newContent } }
-      await expect(client.request(updatePostMutation, updatePostVariables)).resolves.toEqual(
+      const expected = { data: { UpdatePost: { id: 'p9876', content: 'New content' } } }
+      await expect(mutate({ mutation: updatePostMutation, variables })).resolves.toMatchObject(
         expected,
       )
     })
 
-    describe('categories', () => {
-      it('allows a user to update other attributes without passing in categoryIds explicitly', async () => {
-        const expected = { UpdatePost: { id: postId, content: newContent } }
-        await expect(client.request(updatePostMutation, updatePostVariables)).resolves.toEqual(
+    describe('no new category ids provided for update', () => {
+      it('resolves and keeps current categories', async () => {
+        const expected = {
+          data: {
+            UpdatePost: {
+              id: 'p9876',
+              categories: expect.arrayContaining([{ id: 'cat9' }, { id: 'cat4' }, { id: 'cat15' }]),
+            },
+          },
+        }
+        await expect(mutate({ mutation: updatePostMutation, variables })).resolves.toMatchObject(
           expected,
         )
       })
+    })
 
-      it('allows a user to update the categories of a post', async () => {
-        updatePostVariables.categoryIds = ['cat27']
-        await client.request(updatePostMutation, updatePostVariables)
-        const expected = [{ id: 'cat27' }]
-        const postQueryWithCategoriesVariables = {
-          id: postId,
-        }
-        await expect(
-          client.request(postQueryWithCategories, postQueryWithCategoriesVariables),
-        ).resolves.toEqual({ Post: [{ categories: expect.arrayContaining(expected) }] })
+    describe('given category ids', () => {
+      beforeEach(() => {
+        variables = { ...variables, categoryIds: ['cat27'] }
       })
 
-      it('allows a maximum of three category for a successful update', async () => {
-        updatePostVariables.categoryIds = ['cat9', 'cat27', 'cat15', 'cat4']
-        await expect(client.request(updatePostMutation, updatePostVariables)).rejects.toThrow(
-          postSaveError,
+      it('updates categories of a post', async () => {
+        const expected = {
+          data: {
+            UpdatePost: {
+              id: 'p9876',
+              categories: expect.arrayContaining([{ id: 'cat27' }]),
+            },
+          },
+        }
+        await expect(mutate({ mutation: updatePostMutation, variables })).resolves.toMatchObject(
+          expected,
         )
       })
 
+      describe('more than 3 categories', () => {
+        beforeEach(() => {
+          variables = { ...variables, categoryIds: ['cat9', 'cat27', 'cat15', 'cat4'] }
+        })
+
+        it('allows a maximum of three category for a successful update', async () => {
+          const {
+            errors: [error],
+          } = await mutate({ mutation: createPostMutation, variables })
```


should be `updatePostMutation`, no?